File: java/sdk/state/pilot/connected/Releasing.java
package com.fleetmgr.client.sdk.state.pilot.connected;

import com.fleetmgr.client.sdk.event.input.connection.ConnectionEvent;
import com.fleetmgr.client.sdk.event.input.connection.Received;
import com.fleetmgr.client.sdk.event.input.user.UserEvent;
import com.fleetmgr.client.sdk.state.State;
import com.fleetmgr.interfaces.facade.control.ClientMessage;
import com.fleetmgr.interfaces.facade.control.Command;
import com.fleetmgr.interfaces.facade.control.ControlMessage;
import com.fleetmgr.interfaces.facade.control.Response;

/**
 * Created by: Bartosz Nawrot
 * Date: 24.09.2018
 * Description:
 */
public class Releasing extends State {

    Releasing(State state) {
        super(state);
    }

    @Override
    public State start() {
        closeAllChannels();
        send(ClientMessage.newBuilder()
                .setCommand(Command.RELEASE)
                .build());
        return null;
    }

    @Override
    public State notifyConnection(ConnectionEvent event) {
        switch (event.getType()) {
            case RECEIVED:
                return handleMessage(((Received)event).getMessage());

            default:
                return defaultEventHandle(event.toString());
        }
    }

    @Override
    public State notifyEvent(UserEvent event) {
        return defaultEventHandle(event.toString());
    }

    private State handleMessage(ControlMessage message) {
        switch (message.getCommand()) {
            case RELEASE:
                if (message.getResponse() == Response.ACCEPTED) {
                    return new Released(this);

                } else {
                    return defaultMessageHandle(message);
                }

            case HEARTBEAT:
                trace("Heartbeat ignored during release procedure");
                return null;

            default:
                return defaultMessageHandle(message);
        }
    }

    @Override
    public String toString() {
        return "Releasing";
    }
}

File: cpp/source/state/pilot/connected/Releasing.cpp
#include "state/pilot/connected/Releasing.hpp"

#include <state/pilot/connected/Released.hpp>

#include "event/input/connection/Received.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::pilot;
using namespace fm::state::pilot::connected;

using namespace com::fleetmgr::interfaces::facade::control;

using event::input::user::UserEvent;
using event::input::connection::ConnectionEvent;
using event::input::connection::Received;

using event::output::FacadeEvent;

Releasing::Releasing(IState& state) :
    IState(state)
{
}

std::unique_ptr<IState> Releasing::start()
{
    backend.getChannelsHandler().closeAllChannels();
    ClientMessage message;
    message.set_command(Command::RELEASE);
    send(message);
    return nullptr;
}

std::string Releasing::toString() const
{
    return "Releasing";
}

std::unique_ptr<IState> Releasing::handleUserEvent(const UserEvent& event)
{
    switch (event.getType())
    {
    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> Releasing::handleConnectionEvent(const ConnectionEvent& event)
{
    switch (event.getType())
    {
    case ConnectionEvent::RECEIVED:
        return handleMessage(reinterpret_cast<const Received&>(event).getMessage());

    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> Releasing::handleMessage(const ControlMessage& message)
{
    switch (message.command())
    {
    case Command::RELEASE:
        if (message.response() == Response::ACCEPTED)
        {
            return std::make_unique<Released>(*this);
        }
        else
        {
            return defaultMessageHandle(message);
        }
        break;

    default:
        return defaultMessageHandle(message);
    }
}
